eng/common/TestResources/New-TestResources.ps1

```diff
@@ -177,44 +185,46 @@ if ($ProvisionerApplicationId) {
     $provisionerSecret = ConvertTo-SecureString -String $ProvisionerApplicationSecret -AsPlainText -Force
     $provisionerCredential = [System.Management.Automation.PSCredential]::new($ProvisionerApplicationId, $provisionerSecret)
 
-    # Use the given subscription ID if provided.
-    $subscriptionArgs = if ($SubscriptionId) {
-        @{SubscriptionId = $SubscriptionId}
-    }
-    else {
-        @{}
-    }
-
     $provisionerAccount = Retry {
         Connect-AzAccount -Force:$Force -Tenant $TenantId -Credential $provisionerCredential -ServicePrincipal -Environment $Environment @subscriptionArgs
     }
 
     $exitActions += {
         Write-Verbose "Logging out of service principal '$($provisionerAccount.Context.Account)'"
 
-        # Only attempt to disconnect if the -WhatIf flag was not set.  Otherwise, this call is not necessary and will fail.
+        # Only attempt to disconnect if the -WhatIf flag was not set. Otherwise, this call is not necessary and will fail.
         if ($PSCmdlet.ShouldProcess($ProvisionerApplicationId)) {
             $null = Disconnect-AzAccount -AzureContext $provisionerAccount.Context
         }
     }
 }
 elseif (!$CI)
 {
-    # check if user is logged in and login into
-    $context = Get-AzContext;
-    if (!$context)
-    {
-        Log "You are not logged in, connecting to 'Azure SDK Developer Playground'"
-        Connect-AzAccount -Subscription $defaultSubscription
-    }
-    
-    # If no test application id is specified create a new service principal
-    if (!$TestApplicationId) {        
-        Log "TestApplicationId was not specified, creating a new service principal."
+    # If no test application id is specified create a new service principal.
+    if (!$TestApplicationId) {
+
+        # Make sure the user is logged in to create a service principal.
+        $context = Get-AzContext;
+        if (!$context) {
+            Log "You are not logged in; connecting to 'Azure SDK Developer Playground'"
+            Connect-AzAccount -Subscription $defaultSubscription
+        }
+
+        Log "TestApplicationId was not specified; creating a new service principal"
         $servicePrincipal = New-AzADServicePrincipal -Role Owner
 
         $TestApplicationId = $servicePrincipal.ApplicationId
         $TestApplicationSecret = (ConvertFrom-SecureString $servicePrincipal.Secret -AsPlainText);
+
+        # Log in as and run pre- and post-scripts as the service principal.
+        $null = Disable-AzContextAutosave -Scope Process
+
+        Log "Logging into service principal '$TestApplicationId'"
+        $testApplicationCredential = [System.Management.Automation.PSCredential]::new($TestApplicationId, $servicePrincipal.Secret)
+
+        Retry {
+            Connect-AzAccount -Force:$Force -Tenant $context.Tenant.Id -Credential $testApplicationCredential -ServicePrincipal -Environment $Environment @subscriptionArgs
```

Comment:
Why is this under all the ifs. I thought we were planning to always login using the service principal. I think we need to set ProvisionerApplicationId here to TestApplicationId  and let the code inside `if ($ProvisionerApplicationId) {` branch always run.

Reply:
As for why it's in this condition, it's because there's no point in logging in again. This is the only branch where we need to log in as the test application. But, yes, I could see how doing this first and setting the provisioner to the test application could work. We end up redoing a few calls, but compared to the cost of deployment it's nothing substantial.

Reply:
I also wonder if it would make the overall logic more straightforward.

Reply:
Possibly. I do want to keep this file more easily maintainable, which is why I was thinking that I might keep the `!$CI` condition block as-is (easier reading, IMO) unless we want to make this change. If this change we're discussing in this comment helps - and I think it may - it wouldn't really matter anyway.